### ISSUE TITLE
Implement feature extraction in Torch

### ISSUE THREAD

**Issue by deepiksdev** [deepiksdev/torch#6]
See https://github.com/coreylynch/vgg-19-feature-extractor

This extracts a 4000+ dimensions vector. There is an interesting mention about the possibility to find image of a blue car* - "blue" + "red"

**Comment by deepiksdev** [comment]
This will be handled in https://github.com/deepiksdev/microservices/issues/34

**Comment by Lolitawxc** [comment]
@deepiksdev  Isn't 4000 too much ?
Our tensorflow implementation had 2048 features (without PCA) and already took a lot of time.
Doubling the space's dimensionality is going to even more worsen our lookup time (even if we do a PCA afterwards)

**Comment by deepiksdev** [comment]
@Lolitawxc 

>Isn't 4000 too much ?

It's a good question. But can we take less?

**Comment by Lolitawxc** [comment]
@deepiksdev 

> It's a good question. But can we take less?

We could take a different layer or a different pre-trained model.
I have to look at the available pre-trained models in torch and see if one of their last layer is
close to 2k units.

**Comment by deepiksdev** [comment]
@Lolitawxc It would be safe to have the same features as the ones we get on VisualSearch.

**Comment by Lolitawxc** [comment]
@deepiksdev 
> It would be safe to have the same features as the ones we get on VisualSearch.

Then it's vgg16 and 3rd conv layer. Pretrained vgg 16 is available in torch so the issue stays the same:  making feature extraction works.

**Comment by Lolitawxc** [comment]
@deepiksdev 

> Then it's vgg16 and 3rd conv layer.

Sorry I made a mistake. It's obviously inception v3 (but it really was 3rd conv layer).
It's a problem since I believe that no pre-trained inception v3 exists in torch nor caffe.

**Comment by Lolitawxc** [comment]
@deepiksdev 
>  I believe that no pre-trained inception v3 exists in torch nor caffe. 

It's not an official worth but worth trying https://github.com/Moodstocks/inception-v3.torch

**Comment by Lolitawxc** [comment]
@deepiksdev 

I managed to make https://github.com/Moodstocks/inception-v3.torch work so I have a pretrained version of inception v3 in torch.
I am now trying to feature extract the 3rd conv layer.

**Comment by Lolitawxc** [comment]
@deepiksdev I think I have a feature extraction that works.
To be sure that it works I think I am going to feature extract the 30k images from caia, replace the features from VisualSearch with my new features and see if the results are still relevant.
Is that ok with you ?

**Comment by deepiksdev** [comment]
@Lolitawxc 

> To be sure that it works I think I am going to feature extract the 30k images from caia, replace the features from VisualSearch with my new features and see if the results are still relevant.
Is that ok with you ?

Yes, but please keep a backup of the Inception features.

**Comment by Lolitawxc** [comment]
@deepiksdev  For reference, it took 20 min to feature extract 29k images.

**Comment by deepiksdev** [comment]
@Lolitawxc 

> For reference, it took 20 min to feature extract 29k images.

As a reminder, how does this compare with Tensorflow/Inception please?

**Comment by Lolitawxc** [comment]
@deepiksdev 

> As a reminder, how does this compare with Tensorflow/Inception please?

https://github.com/deepiksdev/tensorflow/issues/23#issuecomment-237857224
I remember that it took less than what was predicted, more lire 30/40 min.

**Comment by deepiksdev** [comment]
@Lolitawxc OK, please upload the code to the repository, and document how it works.

**Comment by Lolitawxc** [comment]
@deepiksdev  
> OK, please upload the code to the repository, and document how it works.

Already done.

**Comment by deepiksdev** [comment]
@Lolitawxc 



>> OK, please upload the code to the repository, and document how it works.

> Already done.

There is no link with recent commits in this issue, and no documentation or link with documentation.

**Comment by deepiksdev** [comment]
@Lolitawxc 

> There is no link with recent commits in this issue, and no documentation or link with documentation.

Oh yes, sorry.

**Comment by deepiksdev** [comment]
OK, as agreed, please double check that we obtain similar results as VisualSearch with these features.

**Comment by Lolitawxc** [comment]
@deepiksdev 

> OK, as agreed, please double check that we obtain similar results as VisualSearch with these features.

I am working on it but that's complicated because they don't produce the same numbers even with the same numbers.
So I need to integrate my feature extraction in torch for new images to be compared, since I can't use the Visual search's which uses tensorflow.

**Comment by Lolitawxc** [comment]
@deepiksdev  I don't have all the features copied to Visual search yet only 11k out of 29k but with an image of a swimming pool and an image of a dog the nearest images are deceiving.
So either the nearest images are all in the 18k missing images or I have a preprocessing missing and/or I don't take the good layer.

**Comment by deepiksdev** [comment]
@Lolitawxc Thanks for the update. So, what would you recommend?

**Comment by Lolitawxc** [comment]
@deepiksdev
>So, what would you recommend?

 I compared the pre-processing of inception v3 with my torch feature extraction and tensorflow's. There are indeed different. I am trying to replicate it in torch but I am not sure that I am not missing a step in the pre-processing part.
Regenerating the features to check if I have better results only takes 20 min but scp all the 29k features takes more time.

**Comment by deepiksdev** [comment]
@Lolitawxc OK, thanks

**Comment by Lolitawxc** [comment]
@deepiksdev  My new feature extraction in torch has comparable results to Visual Search's:

![screenshot from 2016-08-10 11 17 20](https://cloud.githubusercontent.com/assets/12127944/17548693/413dcbc2-5eed-11e6-8d12-56d1256852b2.png)


![screenshot from 2016-08-10 11 24 17](https://cloud.githubusercontent.com/assets/12127944/17548689/3c4ab878-5eed-11e6-8fc9-06487a9ca687.png)


I am going to upload the updated code.

**Comment by Lolitawxc** [comment]
@deepiksdev 
When going from 2048 normalized features to 21, the knn stil recognizes duplicate (though fail to recognize similar images):

### NO PCA (2048 normalized features)

![screenshot from 2016-08-10 16 58 04](https://cloud.githubusercontent.com/assets/12127944/17590984/52520762-5fdc-11e6-8b45-15f45d02502f.png)

![screenshot from 2016-08-10 16 56 30](https://cloud.githubusercontent.com/assets/12127944/17590990/55993bb6-5fdc-11e6-979b-a55a2cb9a018.png)

### PCA (21 features)

![screenshot from 2016-08-11 15 50 12](https://cloud.githubusercontent.com/assets/12127944/17590942/2c7a4964-5fdc-11e6-9067-fc41eb19d6a0.png)

![screenshot from 2016-08-11 15 53 45](https://cloud.githubusercontent.com/assets/12127944/17590952/319696c8-5fdc-11e6-81da-1ecd488c0e69.png)

**Comment by deepiksdev** [comment]
@Lolitawxc 

> When going from 2048 normalized features to 21, the knn stil recognizes duplicate (

OK, great. Congrats.

**Comment by deepiksdev** [comment]
No longer relevant
